### ISSUE TITLE
Tooltips on actionable controls - problems with links and in general touch devices

### ISSUE THREAD

**Issue by patrickhlauke** [twbs/bootstrap#23081]
Currently, tooltips have a series of issues:

- if added to a link, it seems that the tooltip swallows any further attempt to actually trigger the link (even when using a mouse or keyboard)
- using a touchscreen, conceptually the tooltip is triggered when the trigger element receives focus; however, if on an interactive element like a button (and link, but that's currently broken in all cases as per the above point), tapping the trigger will already trigger whatever action is associated with it (taking the buttons with tooltips as example, tapping a button will trigger whatever the click behavior associated with it is, and only then show the tooltip)

Ideally, the way tooltips on interactive controls should work:

- for mouse/keyboard, show the tooltips on focus/hover but then do not prevent the default action (not dug into the code, but likely a `preventDefault` somewhere is the culprit)
- for touchscreen/stylus interactions, keep track of whether or not the tooltip is visible; if not, on first tap/focus show tooltip then `preventDefault`. if tooltip IS visible, don't `preventDefault` and allow the control to just carry out its normal function (and possibly close the tooltip). this way interactive controls with tooltip get a "two-step" process: first tap to show tooltip, second tap to actually trigger their functionality

**Comment by patrickhlauke** [comment]
To clarify, I think both issues would be solved by the "prevent default if tooltip is not shown yet, then show tooltip; if already showing, close it and don't prevent default"

**Comment by beardedpayton** [comment]
@patrickhlauke I'm currently looking into this. Just to clarify, are you talking about the tooltip getting hung up on a link after it is clicked. And as link we are talking about an anchor tag correct? Thanks

**Comment by patrickhlauke** [comment]
@beardedpayton no i mean that if an actual link `<a href="foo.com">...</a>` has a tooltip, clicking it doesn't seem to navigate to `foo.com` once the tooltip is shown. the click is swallowed.

and for touch/stylus, the fact that the first tap/touch should trigger the tooltip, but subsequent tap/touch should be passed through (activate the link, trigger the button, etc...basically, `click` needs to transparently go through once tooltip is visible and not be prevented)

**Comment by tyricec** [comment]
Hi, I'm interested into looking into this. Is this issue still relevant?

**Comment by patrickhlauke** [comment]
I believe so, yes.

**Comment by tyricec** [comment]
I've created this codepen, and I seem to be able to click the link and it navigates. Is there something I'm missing in this example?

https://codepen.io/tyricec/pen/VrpBGx

**Comment by patrickhlauke** [comment]
ah, thanks for checking. you're right, some changes to the code since i filed this must have addressed this. good catch.
